lib/vector-core/src/stream/batcher/mod.rs

```diff
@@ -0,0 +1,197 @@
+pub mod config;
+pub mod data;
+pub mod limiter;
+
+use futures::stream::{Fuse, Stream};
+use futures::{Future, StreamExt};
+use pin_project::pin_project;
+
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+use tokio::time::Sleep;
+
+use config::BatchConfig;
+
+#[pin_project]
+pub struct Batcher<S, C> {
+    config: C,
+
+    #[pin]
+    /// The stream this `Batcher` wraps
+    stream: Fuse<S>,
+
+    #[pin]
+    timer: Maybe<Sleep>,
+}
+
+/// An `Option`, but with pin projection
+#[pin_project(project = MaybeProj)]
+pub enum Maybe<T> {
+    Some(#[pin] T),
+    None,
+}
+
+impl<S, C> Batcher<S, C>
+where
+    S: Stream,
+    C: BatchConfig<S::Item>,
+{
+    pub fn new(stream: S, config: C) -> Self {
+        Self {
+            config,
+            stream: stream.fuse(),
+            timer: Maybe::None,
+        }
+    }
+}
+
+impl<S, C> Stream for Batcher<S, C>
+where
+    S: Stream,
+    C: BatchConfig<S::Item>,
+{
+    type Item = C::Batch;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        loop {
+            let mut this = self.as_mut().project();
+            match this.stream.poll_next(cx) {
+                Poll::Ready(None) => {
+                    return {
+                        if this.config.batch_len() == 0 {
+                            Poll::Ready(None)
+                        } else {
+                            Poll::Ready(Some(this.config.take_batch()))
+                        }
+                    }
+                }
+                Poll::Ready(Some(item)) => {
+                    let (item_fits, item_metadata) = this.config.item_fits_in_batch(&item);
+                    if item_fits {
+                        this.config.push_item(item, item_metadata);
+                        if this.config.is_batch_full() {
+                            this.timer.set(Maybe::None);
+                            return Poll::Ready(Some(this.config.take_batch()));
+                        } else if this.config.batch_len() == 1 {
+                            this.timer
+                                .set(Maybe::Some(tokio::time::sleep(this.config.timeout())));
+                        }
+                    } else {
+                        let output = Poll::Ready(Some(this.config.take_batch()));
+                        this.config.push_item(item, item_metadata);
+                        this.timer
+                            .set(Maybe::Some(tokio::time::sleep(this.config.timeout())));
+                        return output;
+                    }
+                }
+                Poll::Pending => {
+                    return {
+                        if let MaybeProj::Some(timer) = this.timer.as_mut().project() {
+                            match timer.poll(cx) {
```

Comment:
You can at least drop one of the if branches by wrapping the timer poll in `ready!(..)` to propagate the pending signal for you.